### PR TITLE
#1: The server responds to GET requests to the root

### PR DESCRIPTION
**Pull request by omarnyte**
This pull request completes Story #1 of the HTTP Server.

## Request
The `Request` class has been refactored with the Builder pattern. All tests that create requests have been updated to create `Request` objects correctly. 

## Response 
I've removed any validation logic from the `Response` class.

## Router 
A router is now instantiated with a hashmap of routes passed in as an argument. That map is currently created in the main `HTTPServer` class.  

## Server
Currently, the `start` method includes the following code: 
``` java        
if (requestString == "") {
  continue;
}
```
Without this code, empty input from the client was being pased to `RequestParser`, which resulted in a NullPointerException when trying to `split` null. I will spend more time addressing this issue in the following story.

### CODE REVIEW

File: src/test/java/Handler/RootHandlerTest.java
Comment (jayden) [inline src/test/java/Handler/RootHandlerTest.java]: This test is failing when I run the tests on my machine. Unit tests should be portable and repeatable - a test should not depend on the environment in which it is running. 

As part of the setup for this test, we can create a temporary directory with some temporary dummy files to use as the path that we will be injecting into our handler. By doing this, we can be certain that we will see the same output no matter where this test is being run. Be sure to delete any temporary data as part of the teardown of the tests.

File: src/test/java/Response/ResponseTest.java
Comment (jayden) [inline src/test/java/Response/ResponseTest.java]: This doesn't seem to be used anywhere in these tests.  I would also rename `goodResponse` as this inaccurately implies that Response performs some validation and determines what's "good" or "bad"

File: src/test/java/Response/StatusLineTest.java
Comment (jayden) [inline src/test/java/Response/StatusLineTest.java]: Why are these tests commented out?

File: src/test/java/ServerTest.java
Comment (jayden) [inline src/test/java/ServerTest.java]: Why are the tests commented out? How do you know that your Server class is working correctly?

File: src/main/java/HTTPServer.java
Comment (jayden) [inline src/main/java/HTTPServer.java]: Consider changing this class name to something else (ex: Main). At a glance, one would erroneously assume that HTTPServer is a subclass of Server given their names.

File: src/main/java/Response/Response.java
Comment (jayden) [inline src/main/java/Response/Response.java]: This doesn't appear to be used anywhere else but the toString() method below

File: src/main/java/Request/Request.java
Comment (jayden) [inline src/main/java/Request/Request.java]: getRequestLine and toString don't appear to be used anywhere.

File: src/main/java/Request/RequestParser.java
Comment (jayden) [inline src/main/java/Request/RequestParser.java]: In idiomatic Java, it's standard convention that private methods are declared below all public methods

**Review by jayden — CHANGES_REQUESTED** [review]
General feedback:

- Rename your commit to something that is more meaningful and communicates the changes that this pull request is providing to the codebase.
- Where is your README? A README is super important in getting new developers/contributors started.
  - No list of required dependencies (gradle/Java 10)
  - No instructions on how to start the server/run the tests
- I'm seeing a lot of magic strings and numbers scattered throughout. Consider wrapping them in meaningfully named constants.

File: src/main/java/HTTPServer.java
Comment (omarnyte) [reply src/main/java/HTTPServer.java]: I actually didn't need that library in this class since I moved the logic out to the Server class, so I've removed it here. But I'll definitely look out for more * imports.

File: src/main/java/HTTPServer.java
Comment (omarnyte) [reply src/main/java/HTTPServer.java]: I've removed the line.

File: src/main/java/Handler/RootHandler.java
Comment (omarnyte) [reply src/main/java/Handler/RootHandler.java]: I've updated the argument name as well as the method names that specified the JAR directory.

File: src/main/java/Request/RequestParser.java
Comment (omarnyte) [reply src/main/java/Request/RequestParser.java]: I've moved `parseRequestLine` into `generateRequest` and reordered the methods.

File: src/main/java/Request/RequestParser.java
Comment (omarnyte) [reply src/main/java/Request/RequestParser.java]: I've fixed that in two different methods.

File: src/main/java/HTTPServer.java
Comment (omarnyte) [reply src/main/java/HTTPServer.java]: I've added exception handling for `ArrayIndexOutOfBoundsException` and `NumberFormatException` on the HTTPServer class. I also added handling for `BindException` and `IllegalArgumentException` in the Server Class.

File: src/main/java/Request/RequestLine.java
Comment (omarnyte) [reply src/main/java/Request/RequestLine.java]: I've added the public modifier key to all constructors.

File: src/main/java/Response/StatusLine.java
Comment (omarnyte) [reply src/main/java/Response/StatusLine.java]: See above.

File: src/main/java/Response/Response.java
Comment (omarnyte) [reply src/main/java/Response/Response.java]: I've integrated the builder pattern, which made it easier to avoid void methods in this class.

File: src/main/java/Handler/NotFoundHandler.java
Comment (omarnyte) [reply src/main/java/Handler/NotFoundHandler.java]: See builder pattern comment below.

File: src/main/java/Handler/RootHandler.java
Comment (omarnyte) [reply src/main/java/Handler/RootHandler.java]: See builder pattern comment below.

File: src/main/java/Response/Response.java
Comment (omarnyte) [reply src/main/java/Response/Response.java]: I've removed this method.

File: src/main/java/Request/Request.java
Comment (omarnyte) [reply src/main/java/Request/Request.java]: I've removed these method.

File: src/main/java/Handler/RootHandler.java
Comment (omarnyte) [reply src/main/java/Handler/RootHandler.java]: I've updated to a for-each loop.

File: src/main/java/Request/Request.java
Comment (omarnyte) [reply src/main/java/Request/Request.java]: Sorry about that! I had commented out the original while I refactored. I've removed it now.

File: src/main/java/Response/StatusLine.java
Comment (omarnyte) [reply src/main/java/Response/StatusLine.java]: I anticipated using it in a similar way as the RequestLine class, but that violates YAGNI, so I've deleted it.

File: src/main/java/Server.java
Comment (omarnyte) [reply src/main/java/Server.java]: I've specified what I'm importing.

File: src/test/java/Handler/RootHandlerTest.java
Comment (omarnyte) [reply src/test/java/Handler/RootHandlerTest.java]: I've configured the setup of the test to create a directory, populate it with two files, run the test, and delete the files and directory. The setup method is currently 25 lines. I'm not sure how to reduce it significantly without the use of void methods to create the directory and each file.

File: src/test/java/Response/ResponseTest.java
Comment (omarnyte) [reply src/test/java/Response/ResponseTest.java]: I've changed the name to `response`.

File: src/test/java/Response/StatusLineTest.java
Comment (omarnyte) [reply src/test/java/Response/StatusLineTest.java]: That was an accident, but I've now deleted the class and its test (see above).

File: src/main/java/HTTPServer.java
Comment (omarnyte) [reply src/main/java/HTTPServer.java]: I've changed the name to `Main`.

File: src/main/java/Request/RequestLine.java
Comment (omarnyte) [reply src/main/java/Request/RequestLine.java]: The numbers represent the three parts of the request line (method, uri, and version). I'm not sure how I'd avoid using numbers in this case. I can probably use regex, but I feel like that would further obscure the meaning.

File: src/main/java/Server.java
Comment (aaizuss) [inline src/main/java/Server.java]: Why did you decide to put the response generation in this class as opposed to the router class? You could have the router choose the proper handler and call generate response instead of just returning a handler. That way the lower level logic is more encapsulated.

File: src/test/java/RouterTest.java
Comment (aaizuss) [inline src/test/java/RouterTest.java]: Generally, we don't like to assert on types because it's more brittle than asserting behavior. I understand why you wrote this because there isn't much else to test here, but you could consider returning a response from the router instead of a specific handler, and then you can test that the response is what you expect instead of just testing the type.

File: src/main/java/Server.java
Comment (omarnyte) [reply src/main/java/Server.java]: I think that the name "Router" suggested to me that the Class was an intermediary that simply sent the request to the Handler. I see how it also makes sense to think about it as an intermediary that sends the response down to the server like a mail delivery person. I like this change. Thank you.

File: src/test/java/RouterTest.java
Comment (aaizuss) [inline src/test/java/RouterTest.java]: test name doesn't quite match what you're testing - rename to `returnsNotFoundWhenURIIsNotFound`

File: src/test/java/RouterTest.java
Comment (aaizuss) [inline src/test/java/RouterTest.java]: There is probably a better way to test the response than creating the exact string (seems like it will be annoying to write these strings as you add routes with complex responses)

Comment (omarnyte) [reply src/test/java/RouterTest.java]: Right now, it feels like my Router tests are redundantly testing the `Response` builder. Is it enough just to test the status code and/or the reason phrase as evidence that the correct handler was returned? If the URI wasn't found, then we should always expect a 404. If the handler was found, but the method was not allowed, we should expect a 405. Successful requests will have a response in the 200's. I'm not sure if some of the upcoming features will require asserting on parts of the response beyond the status code, but--for now--it seems like the status code alone keeps the test at the proper scope.

**Review by jayden — APPROVED** [review]
It's looking a lot better 👍 . I'm gonna approve this to unblock you from submitting your next PR so that you can start rebasing that branch off of master once this has been merged.

**Pull request merged**
